File: packages/java/src/main/java/fbc/Seguranca/Ataques/AtaqueCusum.java
package fbc.Seguranca.Ataques;

import fbc.Seguranca.AlvoCriptografico;
import fbc.Seguranca.AtaqueInterface;
import fbc.Seguranca.ResultadoAtaque;
import fbc.Seguranca.Severidade;
import fbc.Seguranca.SkipAtaqueException;
import fbc.Seguranca.Util;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;

/**
 * Somas cumulativas (cusum), teste do NIST SP800-22. Converte os bits em
 * passos +1/-1 e observa o maior desvio da caminhada aleatória. Um viés
 * pequeno que o monobit quase não vê faz o desvio acumulado crescer.
 * Para uma sequência aleatória, max|S|/sqrt(n) fica tipicamente abaixo de 3.
 */
public class AtaqueCusum implements AtaqueInterface {
    private final int tamanho;
    private final double limiteZ;

    public AtaqueCusum() {
        this(8192, 4.0);
    }

    public AtaqueCusum(int tamanho, double limiteZ) {
        this.tamanho = tamanho;
        this.limiteZ = limiteZ;
    }

    @Override
    public String nome() {
        return "Somas cumulativas (cusum)";
    }

    @Override
    public ResultadoAtaque executar(AlvoCriptografico alvo) {
        byte[] chave = alvo.chaveDeTeste().getBytes(StandardCharsets.UTF_8);
        byte[] ks = alvo.gerarKeystreamBruto(
                chave, Util.randomBytes(alvo.tamanhoIv()), "enc", this.tamanho);
        if (ks == null) {
            throw new SkipAtaqueException("Alvo não expõe gerarKeystreamBruto.");
        }

        int soma = 0;
        int maxAbs = 0;
        for (int i = 0; i < ks.length; i++) {
            int b = ks[i] & 0xFF;
            for (int bit = 7; bit >= 0; bit--) {
                soma += ((b >> bit) & 1) == 1 ? 1 : -1;
                maxAbs = Math.max(maxAbs, Math.abs(soma));
            }
        }
        int n = ks.length * 8;
        double z = maxAbs / Math.sqrt(n);
        boolean vulneravel = z > this.limiteZ;

        Map<String, Object> dados = new HashMap<>();
        dados.put("max_abs", maxAbs);
        dados.put("z", z);

        return new ResultadoAtaque(
                nome(),
                vulneravel,
                vulneravel ? Severidade.MEDIA : Severidade.INFO,
                String.format(
                        Locale.ROOT,
                        "max|S|=%d sobre %d bits, max|S|/sqrt(n)=%.2f (limite=%.1f)",
                        maxAbs,
                        n,
                        z,
                        this.limiteZ),
                dados);
    }
}

File: packages/go/internal/seguranca/ataques/ataque_cusum.go
package ataques

import (
	"fmt"
	"math"

	"criptografia/internal/seguranca"
)

// OpcoesCusum configura o AtaqueCusum.
type OpcoesCusum struct {
	Tamanho int
	LimiteZ float64
}

// AtaqueCusum aplica somas cumulativas (cusum, NIST SP800-22): converte os
// bits em passos +1/-1 e observa o maior desvio da caminhada aleatória.
type AtaqueCusum struct {
	tamanho int
	limiteZ float64
}

// NovoAtaqueCusum cria o ataque. Sem argumento, usa tamanho = 8192 e
// limiteZ = 4.0.
func NovoAtaqueCusum(opcoes ...OpcoesCusum) *AtaqueCusum {
	o := OpcoesCusum{Tamanho: 8192, LimiteZ: 4.0}
	if len(opcoes) > 0 {
		o = opcoes[0]
	}
	return &AtaqueCusum{tamanho: o.Tamanho, limiteZ: o.LimiteZ}
}

// Nome devolve o nome exibido no relatório.
func (a *AtaqueCusum) Nome() string {
	return "Somas cumulativas (cusum)"
}

// Executar roda o ataque.
func (a *AtaqueCusum) Executar(alvo seguranca.AlvoCriptografico) (seguranca.ResultadoAtaque, error) {
	ks := alvo.GerarKeystreamBruto([]byte(alvo.ChaveDeTeste()), seguranca.RandomBytes(alvo.TamanhoIv()), "enc", a.tamanho)

	soma := 0
	maxAbs := 0
	for i := 0; i < len(ks); i++ {
		by := ks[i]
		for b := 7; b >= 0; b-- {
			if ((by >> b) & 1) == 1 {
				soma++
			} else {
				soma--
			}
			if abs := int(math.Abs(float64(soma))); abs > maxAbs {
				maxAbs = abs
			}
		}
	}
	n := len(ks) * 8
	z := float64(maxAbs) / math.Sqrt(float64(n))
	vulneravel := z > a.limiteZ

	severidade := seguranca.SevInfo
	if vulneravel {
		severidade = seguranca.SevMedia
	}

	return seguranca.ResultadoAtaque{
		NomeAtaque: a.Nome(),
		Vulneravel: vulneravel,
		Severidade: severidade,
		Detalhes: fmt.Sprintf(
			"max|S|=%d sobre %d bits, max|S|/sqrt(n)=%.2f (limite=%.1f)",
			maxAbs, n, z, a.limiteZ,
		),
		Dados: map[string]any{"max_abs": maxAbs, "z": z},
	}, nil
}
